controller/controller_server/csi_controller_server.py

```diff
@@ -153,70 +156,69 @@ def CreateVolume(self, request, context):  # pylint: disable=too-many-branches
-            array_mediator.copy_to_existing_volume_from_snapshot(vol_name, src_snapshot_name, src_snapshot_capacity,
-                                                                 min_vol_size, pool)
-            logger.debug("Copy volume from snapshot finished")
-        except controller_errors.VolumeNotFoundError as ex:
-            logger.error("Volume not found while copying snapshot data to volume")
+            source_object = array_mediator.get_object_by_id(source_object_id, source_object_type)
+            if not source_object:
+                self._rollback_create_volume_from_source(array_mediator, volume.id)
+                raise controller_errors.ObjectNotFoundError(source_object_id)
+            source_object_name = source_object.name
+            source_object_capacity = source_object.capacity_bytes
+            logger.debug("Copy {0} {1} data to volume {2}.".format(source_object_type, source_object_id, volume_name))
+            if source_object_type is config.OBJECT_TYPE_NAME_SNAPSHOT:
+                array_mediator.copy_to_existing_volume_from_snapshot(volume_name, source_object_name,
+                                                                     source_object_capacity, minimum_volume_size, pool)
+            else:
+                array_mediator.copy_to_existing_volume_from_volume(volume_name, source_object_name,
+                                                                   source_object_capacity, minimum_volume_size, pool)
+            logger.debug("Copy volume from {0} finished".format(source_object_type))
+        except controller_errors.ObjectNotFoundError as ex:
+            logger.error("Volume not found while copying {0} data to volume".format(source_object_type))
             logger.exception(ex)
+            self._rollback_create_volume_from_source(array_mediator, volume.id)
             raise ex
         except Exception as ex:
-            logger.error("Exception raised while copying snapshot data to volume")
-            self._rollback_create_volume_from_snapshot(array_mediator, vol.id)
+            logger.error("Exception raised while copying {0} data to volume".format(source_object_type))
+            self._rollback_create_volume_from_source(array_mediator, volume.id)
             raise ex
-        return vol
+        return volume
 
     @retry(Exception, tries=5, delay=1)
-    def _rollback_create_volume_from_snapshot(self, array_mediator, vol_id):
-        logger.debug("Rollback copy volume from snapshot. Deleting volume {0}".format(vol_id))
+    def _rollback_create_volume_from_source(self, array_mediator, vol_id):
+        logger.debug("Rollback copy volume from source. Deleting volume {0}".format(vol_id))
         array_mediator.delete_volume(vol_id)
 
-    def _handle_existing_vol_src_snap(self, volume, src_snapshot_id, context):
+    def _handle_existing_vol_src(self, volume, src_object_id, source_type, context):
```

Comment:
this is a good opportunity to rename the short names of this function to full words (e.g. `vol` - `volume`)

Reply:
done